### DISCUSSION TITLE
raise UaStatusCodeError(self.value) asyncua.ua.uaerrors._auto.BadSessionNotActivated: "The session cannot be used because ActivateSession has not been called."(BadSessionNotActivated)

### DISCUSSION THREAD

**Discussion by jeffrey-meyer** [FreeOpcUa/opcua-asyncio#1169] [Q&A]
???

**Comment by AndreasHeine** [comment]
details? 🤣

**Reply by jeffrey-meyer** [reply]
that's the error I get when trying to connect to opc ua server, on a codesys softplc running on a vm

**Reply by AndreasHeine** [reply]
looks like you perform a session transaction before the client has completed to connect!?

usually first socket connect then send hello then secure channel then crate session then activate session then ... read/write ...

https://github.com/FreeOpcUa/opcua-asyncio/blob/d5bacbc8a3542a6037c0352c86088f6587f337fb/asyncua/client/client.py#L269
```python
    async def connect(self):
        """
        High level method
        Connect, create and activate session
        """
        _logger.info("connect")
        await self.connect_socket()
        try:
            await self.send_hello()
            await self.open_secure_channel()
            try:
                await self.create_session()
                try:
                    await self.activate_session(username=self._username, password=self._password, certificate=self.user_certificate)
                except Exception:
                    # clean up session
                    await self.close_session()
                    raise
            except Exception:
                # clean up secure channel
                await self.close_secure_channel()
                raise
        except Exception:
            # clean up open socket
            self.disconnect_socket()
            raise
```

can you provide a code sniped and show how you use the client?

**Comment by YiannisTsitouridis** [comment]
Good day all,
Dear @jeffrey-meyer, I have the exact same problem when trying to connect with a WAGO PLC programmable with Codesys, although I never had this error before with other servers of any kind.
What I initially thought is that the problem occurs when the WAGO device asks the server asks for the user credentials from the client. I understood that when I tried to connect to the WAGO opcua server with the UAExpert Client and a window popped up asking for username and password. However, I haven't found a way to disable this option through Codesys menu yet and to connect only with url.
I am really curious to see if you solved your problem and if @AndreasHeine 's solution was helpful to you.

**Reply by schroeder-** [reply]
You need to set the password user before connecting:
```
client = Client(....)
client.set_user('xxx')
client.set_password('xxx')
async with client:
      ....
```

**Reply by YiannisTsitouridis** [reply]
Thank you very much, actually I was looking for a way to avoid this option though for using in a system with multiple similar servers. 
So I was wondering if there was a way to avoid this option on the server side.
But anyway, I will probably use this one if I cannot avoid credentials.
Again thank you for your help.

**Comment by jmtatsch** [comment]
Have seen the same error when updating python from 3.9 -> 3.11.
It went away when using the latest asyncua==1.0.5 fortunately

**Comment by jmgialamas** [comment]
Good day, everyone.
I too have been receiving this error message however, only when I attempt to use user credentials. Otherwise, I am able to connect and activate session. When I set the credentials- like how @schroeder suggested, and look at Debug level logs- the connect() method will reach await.create_session() and traceback shows the create_session sending from the UASocketProtocol. I get the CreateSessionResponse as well as the find_endpoint INFO log. The difference in the traceback occurs right after the WARNING: asyncua.client.client: Requested session timeout to be 3600000ms, got 100000ms instead. When I use creds the next INFO: is the UAClient calling close_session, when I don't use creds the next INFO: is the UAClient calling activate_session. I can add the actual traceback(s) if needed later...
What confuses me is that I am connected to two other OPC UA servers using creds and they both reach "Sending plain-text password" and connect/disconnect properly.

**Reply by jmgialamas** [reply]
I have a theory as to what is happening related to this documentation, here: https://reference.opcfoundation.org/Core/Part4/v105/docs/5.5

**Reply by AndreasHeine** [reply]
never mind made a mistake myself... deleted the screenshot!

**Reply by AndreasHeine** [reply]
finally i found something which might happen to you... Jesus this was ugly to find!

OK i tested it with a few server configurations:
```ts
config.securityPolicies = [
  SecurityPolicy.None,
  SecurityPolicy.Basic128Rsa15,
  SecurityPolicy.Basic256Sha256,
];
```

the errors only appear with SecurityPolicy.Basic128Rsa15 activated... and its not appearing if i just used SecurityPolicy.Basic256Sha256 exclusively...

![image](https://github.com/user-attachments/assets/cebb0452-8232-4320-a4ea-586fb3e793f5)

there is deftly something with the password encryption...

### TEST1
Server-Config:
```ts
  SecurityPolicy.None,
  SecurityPolicy.Basic128Rsa15,
  SecurityPolicy.Basic256Sha256
```
Client-Output:
```
WARNING:asyncua.client.client:Requested session timeout to be 300000ms, got 3000000ms instead
pw1 http://opcfoundation.org/UA/SecurityPolicy#Basic128Rsa15
http://www.w3.org/2001/04/xmlenc#rsa-1_5
User does not have permission to perform the requested operation.(BadUserAccessDenied)
```
### TEST2
Server-Config:
```ts
  // SecurityPolicy.None,
  // SecurityPolicy.Basic128Rsa15,
  SecurityPolicy.Basic256Sha256
```
Client-Output:
```
WARNING:asyncua.client.client:Requested session timeout to be 300000ms, got 3000000ms instead
pw1 http://opcfoundation.org/UA/SecurityPolicy#Basic256Sha256
http://www.w3.org/2001/04/xmlenc#rsa-oaep
res: [ServerStatusDataType(StartTime=datetime.datetime(2025, 1, 17, 11, 6, 30, 882000, tzinfo=datetime.timezone.utc), CurrentTime=datetime.datetime(2025, 1, 17, 11, 6, 36, 453000, tzinfo=datetime.timezone.utc), State=<ServerState.Running: 0>, BuildInfo_=BuildInfo(ProductUri='SampleServer-productUri', ManufacturerName='SampleServer-manufacturerName', ProductName='SampleServer-productName', SoftwareVersion='node-opcua: 2.140.0', BuildNumber='v1.2.0', BuildDate=datetime.datetime(2024, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)), SecondsTillShutdown=0, ShutdownReason=LocalizedText(Locale=None, Text=None))]
```

between tests client code did not change!

```python
import asyncio
import logging
from asyncua import Client, ua

logging.basicConfig(level=logging.WARNING)
_logger = logging.getLogger('asyncua')

async def main():
    client = Client(url="opc.tcp://127.0.0.1:4840/UA")
    client.set_user("admin")
    client.set_password("pw1")

    try:
        await client.connect()
    except Exception as e: # catching all exceptions is not always wise!
        print(e)
        return

    '''
    clientcode!
    '''
    res = await client.read_values([
        client.get_node("i=2256")
    ])

    print("res:", res)

    await asyncio.sleep(5)

    try:
        await client.disconnect()
    except Exception as e:  # catching all exceptions is not always wise!
        print(e)
        return


if __name__ == "__main__":
    asyncio.run(main())
```


@oroulet @schroeder- any suggestions!?

edit:
![image](https://github.com/user-attachments/assets/fc3e62f7-19e2-4492-a762-35a804666032)

**Reply by jmgialamas** [reply]
Thank you for looking into this, SecurityPolicy.Basic128Rsa15 is activated on this particular server. I can confirm the other servers I was able to connect to using credentials do not have SecurityPolicy.Basic128Rsa15 enabled/available as endpoints.

**Reply by schroeder-** [reply]
Looks like the length field is missing see https://reference.opcfoundation.org/Core/Part4/v104/docs/7.36.2.2#_Ref482553831 . Strange that it works most of the time.

**Reply by AndreasHeine** [reply]
> Looks like the length field is missing see https://reference.opcfoundation.org/Core/Part4/v104/docs/7.36.2.2#_Ref482553831 . Strange that it works most of the time.

i am confused as well!!!

edit:
i guess its an asyncua Client and Server issue because the test did not fail!
https://github.com/FreeOpcUa/opcua-asyncio/blob/master/tests/test_password.py

**Reply by Statkern15** [reply]
Hi!
Are there any updated on how to resolve this issue?
I'm running into the same BadSessionNotActivated issue on one machine. The Basic128Rsa15 security policy is enabled on the server. 
Thanks in advance!

**Reply by schroeder-** [reply]
My best quess is to set a certificate could help.

**Reply by Statkern15** [reply]
Thanks for the fast reply.  Do you mean something like:

```python
await client.set_security(
        SecurityPolicyBasic256Sha256,
        certificate=str(cert),
        private_key=str(private_key),
        server_certificate="certificate-example.der",
    )
```
because i am already doing that ...

**Reply by hasiflo** [reply]
Has anyone solved this issue already? I running into the same problem with asyncua==1.1.8
